libs/parallelism/executors/include/hpx/executors/scheduler_executor.hpp

```diff
@@ -0,0 +1,298 @@
+          : sched_(std::forward<Scheduler>(sched))
+        {
+        }
+
+        /// \cond NOINTERNAL
+        constexpr bool operator==(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ == rhs.sched_;
+        }
+
+        constexpr bool operator!=(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ != rhs.sched_;
+        }
+
+        constexpr auto const& context() const noexcept
+        {
+            return *this;
+        }
+        /// \endcond
+
+        /// \cond NOINTERNAL
+        // Associate the parallel_execution_tag executor tag type as a default
+        // with this executor.
+        using execution_category = parallel_execution_tag;
+
+        // Associate the static_chunk_size executor parameters type as a default
+        // with this executor.
+        using executor_parameters_type = static_chunk_size;
+
+        template <typename T, typename... Ts>
+        using future_type = hpx::future<T>;
+
+        // NonBlockingOneWayExecutor interface
+        template <typename F, typename... Ts>
+        void post(F&& f, Ts&&... ts)
+        {
+            detach(transform(schedule(sched_),
+                hpx::util::deferred_call(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // OneWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) sync_execute(F&& f, Ts&&... ts)
+        {
+            return sync_wait(transform(schedule(sched_),
+                hpx::util::deferred_call(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // TwoWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) async_execute(F&& f, Ts&&... ts)
+        {
+            return make_future(transform(schedule(sched_),
+                hpx::util::deferred_call(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        template <typename F, typename Future, typename... Ts>
+        decltype(auto) then_execute(F&& f, Future&& predecessor, Ts&&... ts)
+        {
+            auto&& predecessor_on_sched =
+                on(keep_future(std::forward<Future>(predecessor)), sched_);
+
+            return make_future(transform(std::move(predecessor_on_sched),
+                hpx::util::bind_back(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // BulkTwoWayExecutor interface
+        template <typename F, typename S, typename... Ts>
+        decltype(auto) bulk_async_execute(F&& f, S const& shape, Ts&&... ts)
+        {
+            using shape_element =
+                typename hpx::traits::range_traits<S>::value_type;
+            using result_type = hpx::util::detail::invoke_deferred_result_t<F,
+                shape_element, Ts...>;
+
+            using size_type = decltype(hpx::util::size(shape));
+            size_type const n = hpx::util::size(shape);
```

Comment:
`n` is needed for the non-void case only.

Reply:
Good catch, fixed.